java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java

```diff
@@ -0,0 +1,272 @@
+  private Future<?> scheduledFuture;
+
+  //TODO: bounded buffer?
+  private List<String> buffer = new ArrayList<>();
+
+  public WavefrontDirectSender(String server, String token) {
+    this.server = server;
+    this.token = token;
+    scheduler = Executors.newScheduledThreadPool(1, new NamedThreadFactory(DEFAULT_SOURCE));
+  }
+
+  private DataIngestorAPI createWavefrontService(String server, String token) {
+    httpClient = HttpClientBuilder.create().
+        setUserAgent(DEFAULT_SOURCE).
+        setRequestExecutor(new TokenHeaderHttpRequestExectutor(token)).
+        build();
+
+    final ApacheHttpClient4Engine apacheHttpClient4Engine = new ApacheHttpClient4Engine(httpClient, true);
+    apacheHttpClient4Engine.setFileUploadInMemoryThresholdLimit(100);
+    apacheHttpClient4Engine.setFileUploadMemoryUnit(ApacheHttpClient4Engine.MemoryUnit.MB);
+
+    ResteasyProviderFactory factory = ResteasyProviderFactory.getInstance();
+    ClientHttpEngine httpEngine = apacheHttpClient4Engine;
+
+    ResteasyClient client = new ResteasyClientBuilder().
+        httpEngine(httpEngine).
+        providerFactory(factory).
+        register(GZIPDecodingInterceptor.class).
+        register(AcceptEncodingGZIPFilter.class).
+        build();
+    ResteasyWebTarget target = client.target(server);
+    return target.proxy(DataIngestorAPI.class);
+  }
+
+  @Override
+  public synchronized void connect() throws IllegalStateException, IOException {
+    if (httpClient == null) {
+      directService = createWavefrontService(server, token);
+      scheduledFuture = scheduler.scheduleAtFixedRate(this, 1, 1, TimeUnit.SECONDS);
+    }
+  }
+
+  @Override
+  public void send(String name, double value) throws IOException {
+    addPoint(name, value, null, DEFAULT_SOURCE, null);
+  }
+
+  @Override
+  public void send(String name, double value, @Nullable Long timestamp) throws IOException {
+    addPoint(name, value, timestamp, DEFAULT_SOURCE, null);
+  }
+
+  @Override
+  public void send(String name, double value, @Nullable Long timestamp, String source) throws IOException {
+    addPoint(name, value, timestamp, source, null);
+  }
+
+  @Override
+  public void send(String name, double value, String source, @Nullable Map<String, String> pointTags) throws IOException {
+    addPoint(name, value, null, source, pointTags);
+  }
+
+  @Override
+  public void send(String name, double value, @Nullable Long timestamp, String source,
+                   @Nullable Map<String, String> pointTags) throws IOException {
+    addPoint(name, value, timestamp, source, pointTags);
+  }
+
+  private void addPoint(@NotNull String name, double value, @Nullable Long timestamp, @NotNull String source,
+                        @Nullable Map<String, String> pointTags) throws IOException {
+
+    //TODO: extra validation required?
+    String point = pointToString(name, value, timestamp, source, pointTags);
+    if (point != null) {
+      synchronized(pointsMutex) {
```

Comment:
This is sort of a bottleneck. But I don't think we are sending high-frequency items via direct ingestion API so bottleneck should not be that bad.
Also, all the points sent to the in-memory buffer via direct ingestion will be lost if the client process is restarted.

Reply:
Use a ReadWriteLock and just acquire read locks here and write lock when you flush.